design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+
+In addition to new `CREATE` properties, the user will also be able to set default key and value 
+formats using system configuration. 
+
+`KEY_FORMAT` will not be a required property _if_ `ksql.persistence.default.format.key` is set. 
+`VALUE_FORMAT` is currently a required property, but will no longer be required _if_ 
+`ksql.persistence.default.format.value` is set.
+
+Somewhat unrelated, but small, we propose also changing the `KAFKA_TOPIC` property from _required_ to 
+_optional_ in `CREATE STREAM` and `CREATE TABLE` statements. If not supplied, it will default to the 
+uppercase name of the stream or table.  This matches the behaviour of the property in 
+`CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements, where it is already optional.
+
+To additional configurations will be added to control topic names:
+
+* `ksql.persistence.topic.name.case`: if `lower`, defaults ksql to using lowercase topics names
+  where topic names are not explicitly provided. This includes matching a `CREATE TABLE` or 
+  `CREATE STREAM` statement to an existing topic. Default: `upper`.
+* `ksql.persistence.topic.name.case.insensitive`: if `true`, matching a `CREATE TABLE` or 
+  `CREATE STREAM` statement to an existing topic will be case insensitive. Multiple matches will 
+  result in an error. Default: `false`.
+
+With these changes there will no longer be any required `CREATE` properties. This means the `WITH` 
+clause for `CREATE` statements for streams and tables would be optional. We propose this makes the 
+syntax more intuitive for those already familiar with SQL. This would make a `CREATE TABLE` 
+statement ANSI standard.
+
+```sql
+SET 'ksql.persistence.default.format.key'='Avro';
+SET 'ksql.persistence.default.format.value'='Avro';
+
+CREATE TABLE USERS (
+  ID BIGINT PRIMARY KEY,
+  NAME STRING
+);
+
+CREATE STREAM BIDS (
+  ITEM_ID BIGINT KEY,
+  USER_ID BIGINT,
+  AMMOUNT INT
+);
+```
+
+### Implementation
+
+The system already serializes the key format of source, intermediate and sink topics as part of the
+query plan, meaning it should be fairly easily to plug in new formats. 
+
+Validation will be added to ensure only supported key formats are set, and that key column data types
+are supported by key formats.
+
+Most existing functionality should _just work_, as the key format only comes into play during 
+(de)serialization, (obviously). The only area where additional work is expected is joins.
+
+Joins require the binary key of both sides of the join to match and both sides to be delivered to 
+the same ksqlDB node.  The former normally ensuring the latter, unless a custom partitioning 
+strategy has been used.
+
+The introduction of additional key formats means that while the deserialized key from both sides of 
+a join may match, the serialized binary data may differ if the key serialization format is different.
+To accommodate this, ksqlDB will automagically repartition the right side of a join to match the key
```

Comment:
Why the right side?

Reply:
Because it's better than the wrong side, obviously!

Unlike a traditional system, we can't easily _know_ which side is the cheapest to repartition.  When queries are started the topics in question could be empty.  Even if they aren't empty, we can't really read anything into the size of the topic. A topic may be huge, while the other side tiny, but that's because the huge topic contains a pretty static table, and the other side is about to have a fire hose turned on. We can't even assume a large change log means a large table, it could just be bad/no compaction.

Without an optimiser to choose which side to repartition, I figured we just have to pick one.  And it felt more likely that the right would be the cheaper, on average, to repartition, mainly because of stream-table joins, where the table doesn't trigger joins and will often be updated less frequently than the right.

The user is always free to explicitly repartition themselves if they don't like our choice.

Of course, this is just a gut feeling. So feel free to plow in and brake all my assumptions ;)

Reply:
I've extended the detail about how we will pick the formats during joins, though the actual algo is still not defined: https://github.com/confluentinc/ksql/blob/dcd28ff1b26705625a477db4e3e81a1891b184aa/design-proposals/klip-33-key-format.md#implementation